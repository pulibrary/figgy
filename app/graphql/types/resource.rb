# frozen_string_literal: true
module Types::Resource
  include Types::BaseInterface
  description "A resource in the system."
  orphan_types Types::ScannedResourceType, Types::FileSetType, Types::ScannedMapType

  field :id, String, null: true
  field :label, String, null: true
  field :viewing_hint, String, null: true
  field :url, String, null: true
  field :members, [Types::Resource], null: true
  field :source_metadata_identifier, String, null: true
  field :thumbnail, Types::Thumbnail, null: true

  definition_methods do
    def resolve_type(object, _context)
      "Types::#{object.class}Type".constantize
    end
  end

  def members
    @members ||= Wayfinder.for(object).members
  end

  def manifest_url
    helper.polymorphic_url([:manifest, object])
  end

  def url
    @url ||= helper.show_url(object)
  end

  # We need to centralize logic for navigating a MVW's members to find a
  # thumbnail file set. This is a hack to use the helper's logic for doing that.
  # refactor ticketed as https://github.com/pulibrary/figgy/issues/1708
  def helper
    @helper ||= ManifestBuilder::ManifestHelper.new.tap do |helper|
      helper.singleton_class.include(ThumbnailHelper)
      helper.define_singleton_method(:image_tag) do |url, _opts|
        url
      end
      helper.define_singleton_method(:image_path) do |url|
        url
      end
    end
  end

  def thumbnail
    return if object.try(:thumbnail_id).blank? || thumbnail_resource.blank?
    {
      id: thumbnail_resource.id.to_s,
      thumbnail_url: helper.figgy_thumbnail_path(thumbnail_resource),
      iiif_service_url: helper.figgy_thumbnail_path(thumbnail_resource).gsub("/full/!200,150/0/default.jpg", "")
    }
  end

  def thumbnail_resource
    @thumbnail_resource ||= query_service.find_by(id: object.try(:thumbnail_id).first)
  rescue Valkyrie::Persistence::ObjectNotFoundError
    nil
  end

  def query_service
    Valkyrie::MetadataAdapter.find(:indexing_persister).query_service
  end
end
