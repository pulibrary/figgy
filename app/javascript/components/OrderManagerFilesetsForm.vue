<template>
  <div>
    <heading level="h2">
      Generate Labels <small>for selected items</small>
    </heading>
    <form
      id="app"
      novalidate="true"
    >
      <input-text
        id="unitLabel"
        v-model="unitLabel"
        label="Label"
        placeholder="e.g., p."
        @input="updateMultiLabels()"
      />
      <input-text
        id="startNum"
        v-model="start"
        label="Starting Numeral"
        placeholder="e.g., 10"
        @input="updateMultiLabels()"
      />
      <input-checkbox
        v-if="!isMultiVolume"
        v-model="bracket"
        :options="addBracketOpts"
        @change="updateMultiLabels()"
      />

      <input-select
        v-if="!isMultiVolume"
        id="labelMethod"
        v-model="method"
        label="Labeling Method"
        :options="methodOpts"
      />

      <div
        v-if="method === 'foliate'"
        class="lux-row"
      >
        <input-text
          id="frontLabel"
          v-model="frontLabel"
          label="Front Label"
          placeholder="(recto)"
          @input="updateMultiLabels()"
        />
        <input-text
          id="backLabel"
          v-model="backLabel"
          label="Back Label"
          placeholder="(verso)"
          @input="updateMultiLabels()"
        />
        <input-select
          v-if="!isMultiVolume"
          id="startWith"
          v-model="startWith"
          label="Start With"
          :options="startWithOpts"
          @change="updateMultiLabels()"
        />
      </div>
    </form>
  </div>
</template>

<script>
// import Lablr from "../utils/lablr"
// const Lablr = require('../utils/lablr').default
import Lablr from 'page-label-generator'
import { mapState } from 'vuex'
/**
 * This is the Filesets Form for the Order Manager in Figgy
 */
export default {
  name: 'FilesetsForm',
  status: 'ready',
  release: '1.0.0',
  type: 'Pattern',
  metaInfo: {
    title: 'Fileset Form',
    htmlAttrs: {
      lang: 'en'
    }
  },
  props: {
    /**
     * The html element name used for the component.
     */
    type: {
      type: String,
      default: 'div'
    }
  },
  data: function () {
    return {
      start: '1',
      method: 'paginate',
      frontLabel: 'r. ',
      backLabel: 'v. ',
      startsWith: 'front',
      unitLabel: 'p. ',
      bracket: false
    }
  },
  computed: {
    ...mapState({
      resource: state => state.ordermanager.resource,
      gallery: state => state.gallery
    }),
    isMultiVolume () {
      return this.$store.getters.isMultiVolume
    },
    selectedTotal () {
      return this.gallery.selected.length
    },
    addBracketOpts: function () {
      return [
        {
          name: 'addBrackets',
          value: 'Add Brackets',
          id: 'addBrackets',
          checked: this.labelerOpts().bracket
        }
      ]
    },
    methodOpts: function () {
      return [{ label: 'Paginate (Default)', value: 'paginate' }, { label: 'Foliate', value: 'foliate' }]
    },
    startWithOpts: function () {
      return [{ label: 'Front (Default)', value: 'front' }, { label: 'Back', value: 'back' }]
    }
  },
  watch: {
    method: function (val) {
      this.updateMultiLabels()
    }
  },
  methods: {
    labelerOpts () {
      let unitLabel = this.unitLabel

      // This should be generated with calculate() or watch()
      if (this.method === 'paginate') {
        unitLabel = 'p. '
      } else if (this.method === 'foliate') {
        unitLabel = 'f. '
      }

      let frontLabel = this.method === 'paginate' ? '' : this.frontLabel
      let backLabel = this.method === 'paginate' ? '' : this.backLabel

      return {
        start: this.start,
        method: this.method,
        startsWith: this.startsWith,
        bracket: this.bracket,
        frontLabel,
        backLabel,
        unitLabel
      }
    },
    isNormalInteger (str) {
      return /^\+?(0|[1-9]\d*)$/.test(str)
    },
    updateMultiLabels () {
      let changeList = this.gallery.changeList
      let items = this.gallery.items
      this.start = this.isNormalInteger(this.start)
        ? this.start - 0
        : this.start
      let generator = Lablr.pageLabelGenerator(this.labelerOpts())
      for (let i = 0; i < this.selectedTotal; i++) {
        let index = this.gallery.items
          .map(function (item) {
            return item.id
          })
          .indexOf(this.gallery.selected[i].id)
        items[index].caption = generator.next().value

        if (changeList.indexOf(this.gallery.selected[i].id) === -1) {
          changeList.push(this.gallery.selected[i].id)
        }
      }

      this.$store.dispatch('updateChanges', changeList)
      this.$store.dispatch('updateItems', items)
    }
  }
}
</script>

<style lang="scss" scoped>
small {
  font-size: 1rem;
  font-weight: 400;
}
</style>
