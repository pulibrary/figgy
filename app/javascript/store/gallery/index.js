export const galleryState = {
  items: [],
  selected: [],
  cut: [],
  changeList: [],
  ogItems: [],
}

export const galleryMutations = {
  CUT(state, itemArray) {
    state.cut = [...itemArray]
  },
  PASTE(state, itemArray) {
    state.items = [...itemArray]
  },
  SELECT(state, itemArray) {
    state.selected = [...itemArray]
  },
  SET_GALLERY(state, items) {
    state.items = items
    state.ogItems = items
  },
  SORT_ITEMS(state, value) {
    state.items = [...value]
  },
  UPDATE_CHANGES(state, changeList) {
    state.changeList = [...changeList]
  },
  UPDATE_ITEMS(state, items) {
    state.items = [...items]
  },
}

export const galleryModule = {
  state: galleryState,
  mutations: galleryMutations,
}
