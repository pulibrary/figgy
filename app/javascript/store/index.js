import Vue from 'vue/dist/vue.esm'
import Vuex from 'vuex'
import actions from './vuex/actions'
import {modules} from 'lux-design-system'
Vue.use(Vuex)

const store = new Vuex.Store({
  actions,
  modules: {
    ordermanager: modules.resourceModule,
    gallery: modules.galleryModule,
  }
})

export default store
