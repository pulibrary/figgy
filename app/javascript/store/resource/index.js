export const resourceState = {
  resource: {
    id: "",
    resourceClassName: "",
    bibId: "",
    label: "Resource not available.",
    thumbnail: "",
    startCanvas: "",
    isMultiVolume: false,
    viewingHint: null,
    viewingDirection: null,
    members: [],
    loadState: "NOT_LOADED",
    saveState: "NOT_SAVED",
    ogState: {},
  },
}

export const resourceMutations = {
  APPLY_STATE(state) {
    state.gallery.ogItems = state.gallery.items
    state.gallery.changeList = []
    state.resource.ogState = {
      startCanvas: resource.startPage,
      thumbnail: resource.thumbnail,
      viewingHint: resource.viewingHint,
      viewingDirection: resource.viewingDirection,
    }
    state.resource.saveState = "NOT_SAVED"
  },
  CHANGE_RESOURCE_LOAD_STATE(state, loadState) {
    state.resource.loadState = loadState
  },
  SAVED_STATE(state, saveStatus) {
    state.resource.saveState = saveStatus
  },
  SET_RESOURCE(state, resource) {
    state.resource.id = resource.id
    state.resource.resourceClassName = resource.__typename
    state.resource.label = resource.label
    state.resource.startCanvas = resource.startPage
    state.resource.viewingHint = resource.viewingHint
    state.resource.viewingDirection = resource.viewingDirection
    state.resource.thumbnail = resource.thumbnail != null ? resource.thumbnail.id : null
    state.resource.members = resource.members

    const items = resource.members.map(member => ({
      id: member.id,
      viewingHint: member.viewingHint != null ? member.viewingHint : "single",
      caption: member.label, // member.__typename + " : " + member.id,
      service:
        member["thumbnail"] && typeof(member.thumbnail.iiifServiceUrl) != "undefined"
          ? member.thumbnail.iiifServiceUrl
          : Global.figgy.resource.defaultThumbnail,
      mediaUrl:
        member["thumbnail"] && typeof(member.thumbnail.iiifServiceUrl) != "undefined"
          ? member.thumbnail.iiifServiceUrl + "/full/300,/0/default.jpg"
          : Global.figgy.resource.defaultThumbnail,
    }))
    state.gallery.items = items
    state.gallery.ogItems = items
    state.resource.loadState = "LOADED"
    state.resource.ogState = {
      startCanvas: resource.startPage,
      thumbnail: resource.thumbnail != null ? resource.thumbnail.id : null,
      viewingHint: resource.viewingHint,
      viewingDirection: resource.viewingDirection,
    }
  },
  UPDATE_STARTCANVAS(state, startCanvas) {
    state.resource.startCanvas = startCanvas
  },
  UPDATE_THUMBNAIL(state, thumbnail) {
    state.resource.thumbnail = thumbnail
  },
  UPDATE_VIEWDIR(state, viewDir) {
    state.resource.viewingDirection = viewDir
  },
  UPDATE_VIEWHINT(state, viewHint) {
    state.resource.viewingHint = viewHint
  },
}

export const resourceGetters = {
  getMemberCount: state => {
    return state.resource.members.length
  },
  isMultiVolume: state => {
    const volumes = state.resource.members.filter(member => member.__typename === "ScannedResource")
    return volumes.length > 0 ? true : false
  },
  orderChanged: state => {
    let ogOrder = JSON.stringify(state.gallery.ogItems.map(item => item.id))
    let imgOrder = JSON.stringify(state.gallery.items.map(item => item.id))
    return ogOrder !== imgOrder
  },
  stateChanged: (state, getters) => {
    var propsChanged = []
    propsChanged.push(state.resource.ogState.thumbnail !== state.resource.thumbnail)
    propsChanged.push(state.resource.ogState.startCanvas !== state.resource.startCanvas)
    propsChanged.push(state.resource.ogState.viewingHint !== state.resource.viewingHint)
    propsChanged.push(state.resource.ogState.viewingDirection !== state.resource.viewingDirection)
    propsChanged.push(state.gallery.changeList.length > 0)
    propsChanged.push(getters.orderChanged)
    if (propsChanged.indexOf(true) > -1) {
      return true
    } else {
      return false
    }
  },
}
