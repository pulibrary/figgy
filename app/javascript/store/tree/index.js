export const treeState = {
  selected: null,
}

export const treeMutations = {
  SELECT(state, selectedId) {
    state.selected = selectedId
  },
}
