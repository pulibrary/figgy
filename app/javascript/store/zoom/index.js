export const zoomState = {
  zoomed: null,
}

export const zoomMutations = {
  ZOOM(state, obj) {
    state.zoomed = obj
  },
  RESET_ZOOM(state) {
    state.zoomed = null
  },
}

export const zoomGetters = {
  getIsZoomed: state => {
    return state.zoomed
  },
}
