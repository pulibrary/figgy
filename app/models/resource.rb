# frozen_string_literal: true
class Resource < Valkyrie::Resource
  def self.human_readable_type
    default = @_human_readable_type || name.demodulize.titleize
    I18n.translate("models.#{new.model_name.i18n_key}", default: default)
  end

  def self.can_have_manifests?
    false
  end

  def human_readable_type
    self.class.human_readable_type
  end

  def self.model_name
    ::ActiveModel::Name.new(self)
  end

  # Determines if this is an image resource
  # @return [TrueClass, FalseClass]
  def image_resource?
    false
  end

  # Determines if this is a geospatial resource
  # @return [TrueClass, FalseClass]
  def geo_resource?
    false
  end

  # Determines if this is a media resource
  # @return [TrueClass, FalseClass]
  def media_resource?
    false
  end

  # Virtual property used for stashing pre-loaded objects. Populated by a query.
  attr_accessor :loaded
  def loaded
    @loaded ||= {}
  end
end
