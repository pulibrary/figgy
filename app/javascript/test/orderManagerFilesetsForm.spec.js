import Vuex from "vuex"
import { createLocalVue, mount, shallowMount } from "@vue/test-utils"
import OrderManagerFilesetsForm from "../components/OrderManagerFilesetsForm.vue"
import { resourceMutations, resourceGetters } from "../store/resource"
import { modules } from 'lux-design-system'

// create an extended `Vue` constructor
const localVue = createLocalVue()
localVue.use(Vuex)

let wrapper
let getters
let options
let store
let actions
let items = [
  { id: "a", title: "A", viewingHint: "single", caption: "a", mediaUrl: "https://picsum.photos/600/300/?random" },
  { id: "b", title: "B", caption: "b", mediaUrl: "https://picsum.photos/600/300/?random" },
  { id: "c", title: "C", caption: "c", mediaUrl: "https://picsum.photos/600/300/?random" },
]

describe("OrderManagerFilesetsForm.vue", () => {
  beforeEach(() => {
    actions = {
      updateItems: jest.fn(),
      updateChanges: jest.fn()
    }

    const gallery = {
      state: {
        items: items,
        selected: [items[0]],
        cut: [],
        changeList: [],
        ogItems: items,
      },
      mutations: modules.galleryModule.mutations,
    }

    const getters = {
      isMultiVolume: () => false
    }

    const resource = {
      state: {
        resource: {
          id: "",
          resourceClassName: "",
          bibId: "",
          label: "Resource not available.",
          thumbnail: "",
          startCanvas: "",
          isMultiVolume: false,
          viewingHint: "single",
          viewingDirection: null,
          members: [],
          loadState: "NOT_LOADED",
          saveState: "NOT_SAVED",
          ogState: {},
        },
      },
      mutations: resourceMutations,
      getters: getters,
      actions: actions,
      modules: {
        gallery: gallery,
      },
    }

    store = new Vuex.Store({
      modules: {
        ordermanager: resource,
        gallery: gallery,
      },
    })

    options = {
      computed: {
        orderChanged: function () {
          return getters.orderChanged
        },
        isDisabled: function () {
          if (getters.stateChanged) {
            return false
          } else {
            return true
          }
        }
      }
    }

    wrapper = mount(OrderManagerFilesetsForm, {
      options,
      localVue,
      store,
      stubs: ["heading","input-select", "input-text", "input-checkbox"],
    })
  })

  it('allows the update of multiple labels', () => {
    wrapper.vm.updateMultiLabels()
    expect(actions.updateChanges).toHaveBeenCalled()
    expect(actions.updateItems).toHaveBeenCalled()
  })

  it('updates default label when switching pagination method', () => {
    expect(wrapper.vm.labelerOpts().unitLabel).toEqual('p. ')

    wrapper.vm.method = 'foliate'
    expect(wrapper.vm.labelerOpts().unitLabel).toEqual('f. ')

    wrapper.vm.method = 'paginate'
    expect(wrapper.vm.labelerOpts().unitLabel).toEqual('p. ')
  })

  it('does not set frontLabel/backLabel by default', () => {
    expect(wrapper.vm.labelerOpts().frontLabel).toEqual('')
    expect(wrapper.vm.labelerOpts().backLabel).toEqual('')

    wrapper.vm.method = 'foliate'
    expect(wrapper.vm.labelerOpts().frontLabel).toEqual('r. ')
    expect(wrapper.vm.labelerOpts().backLabel).toEqual('v. ')

    wrapper.vm.method = 'paginate'
    expect(wrapper.vm.labelerOpts().frontLabel).toEqual('')
    expect(wrapper.vm.labelerOpts().backLabel).toEqual('')
  })


  it("has the expected html structure", () => {
    expect(wrapper.element).toMatchSnapshot()
  })
})
