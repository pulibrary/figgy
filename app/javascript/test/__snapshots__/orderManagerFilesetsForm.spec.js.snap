// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OrderManagerFilesetsForm.vue has the expected html structure 1`] = `
<div>
  <heading-stub
    level="h2"
  >
    
    Generate Labels 
    <small>
      for selected items
    </small>
  </heading-stub>
   
  <form
    id="app"
    novalidate="novalidate"
  >
    <input-text-stub
      id="unitLabel"
      label="Label"
      placeholder="e.g., p."
      value="p. "
    />
     
    <input-text-stub
      id="startNum"
      label="Starting Numeral"
      placeholder="e.g., 10"
      value="1"
    />
     
    <input-checkbox-stub
      options="[object Object]"
    />
     
    <input-select-stub
      id="labelMethod"
      label="Labeling Method"
      options="[object Object],[object Object]"
      value="paginate"
    />
     
    <!---->
  </form>
</div>
`;
